src/fping.c

```diff
@@ -2153,14 +2161,16 @@ int decode_icmp_ipv4(
     char *reply_buf,
     size_t reply_buf_len,
     unsigned short *id,
-    unsigned short *seq)
+    unsigned short *seq,
+    int *ip_header_tos)
 {
     struct icmp *icp;
+    struct ip *ip;
     int hlen = 0;
 
-    if (!using_sock_dgram4) {
-        struct ip *ip = (struct ip *)reply_buf;
+    ip = (struct ip *)reply_buf;
```

Comment:
I do not think this is correct. It seems to me that only a raw socket provides the IP header as part of the received data. The code after the `if (…) {…}` assumes that the ICMP datagram starts directly at the beginning of `recv_buf` when using a datagram socket.

Reply:
In my opinion, the code is correct.
Both sockets return an ip header with the TOS value.

* Type Of Service == TOS

**SOCK_RAW**
```
IP Header:
  |-Version        : 4
  |-Internet Header Length  : 5 DWORDS or 20 Bytes
  |-Type Of Service   : 0
  |-Total Length   : 84  Bytes
  |-Identification  : 58676
  |-Time To Live    : 64
  |-Protocol     : 1
  |-Header Checksum : 22386
  |-Source IP       : 127.0.0.1
  |-Destination IP  : 127.0.0.1
```

**SOCK_DGRAM**
```
IP Header:
  |-Version        : 7
  |-Internet Header Length  : 4 DWORDS or 16 Bytes
  |-Type Of Service   : 101
  |-Total Length   : 29556  Bytes
  |-Identification  : 2560
  |-Time To Live    : 0
  |-Protocol     : 0
  |-Header Checksum : 0
  |-Source IP       : 0.0.0.0
  |-Destination IP  : 0.0.0.0
```

I checked this in advance in a separate program

Reply:
This shows that the data received from a `SOCK_DGRAM` socket does *not* start with an IP header:
 * The valid versions for IP headers are 4 or 6, not 7. Since we are handling ICMP via IPv4 the version must be 4.
 * The minimum IPv4 header length is 5 (20 bytes), not 4 (16 bytes)
 * The protocol is given as 0, but the protocol number for ICMP for IPv4 is 1

The code finds the ICMP data at offset `hlen` from the start of the data `reply_buf`:
```
icp = (struct icmp *)(reply_buf + hlen);
```
In case of `using_sock_dgram4`, `hlen` equals 0, and the ICMP data starts at the beginning of `reply_buf`. In case of `!using_sock_dgram4`, the value of `hlen` is extracted from the IP header at the start of `reply_buf` in order to skip it to get at the ICMP data.

The data can either start with an IPv4 header followed by ICMP, or start with ICMP, but not both at the same time.

Reply:
You could do the following:
* in `decode_icmp_ipv4()`:
  * initialize the TOS value to -1 (`*ip_header_tos = -1;`)
  * set it to the received value inside the `if (!using_sock_dgram4) {…}` block
* in `wait_for_reply()`:
  * only print the received TOS value if it is greater than -1 (you could print `(TOS unknown)` or just not add the output)